src/Polly.Core/Utils/TaskHelper.cs

```diff
@@ -34,4 +36,58 @@ public static TResult GetResult<TResult>(this ValueTask<TResult> task)
 
         return task.Preserve().GetAwaiter().GetResult();
     }
+
+    /// <summary>
+    /// Convert a <see cref="ValueTask{T}"/> to a <see cref="ValueTask{Outcome{T}}"/>, transfering the potential
+    /// exceptions into the <see cref="Outcome{T}"/> to avoid rethrowing them when the task is awaited.
+    /// </summary>
+    /// <param name="task">Task to convert.</param>
+    /// <typeparam name="T">Outcome value.</typeparam>
+    /// <returns>An successful or unsuccessful outcome.</returns>
+    public static ValueTask<Outcome<T>> ToAsyncOutcome<T>(this ValueTask<T> task) => ToAsyncOutcomeAwaiter<T>.GetTaskOutcome(task);
+
+    private readonly struct ToAsyncOutcomeAwaiter<T>
+    {
+        private readonly ValueTask<T> _task;
+        private readonly TaskCompletionSource<Outcome<T>> _tcs;
+
+        private ValueTask<Outcome<T>> TaskResult => new(_tcs.Task);
+
+        public static ValueTask<Outcome<T>> GetTaskOutcome(ValueTask<T> task) =>
+            new ToAsyncOutcomeAwaiter<T>(task).TaskResult;
+
+        private ToAsyncOutcomeAwaiter(ValueTask<T> task)
+        {
+            _task = task;
+            _tcs = new TaskCompletionSource<Outcome<T>>();
+
+            // Setup task completed callback
+            ValueTaskAwaiter<T> awaiter = _task.GetAwaiter();
```

Comment:
What happens when there is an exception at the beginning of async callback? (before the yield statement)